File: memory/copy/struct.go
// go run struct.go
package main

import "fmt"

type Employee struct {
	Id     int
	Name   string // immutable utf-8 encoded
	Stats  []int
	Jersey [1]int
}

func deep_copy(e Employee) Employee {
	c := Employee{
		Id:     e.Id,
		Name:   e.Name,
		Stats:  make([]int, len(e.Stats)),
		Jersey: e.Jersey,
	}
	copy(c.Stats, e.Stats)
	return c
}

func main() {
	// initialize struct
	e1 := Employee{
		Id:     3,
		Name:   "freeman", // just a pointer to string
		Stats:  []int{3, 4, 5},
		Jersey: [1]int{5}, // deep copy
	}

	// deep copy
	e2 := deep_copy(e1)

	// shallow copy
	e3 := e1

	// update
	e1.Id = 1
	e1.Name = "ohtani"
	e1.Stats[0] = 1
	e1.Stats[1] = 2
	e1.Stats[2] = 3
	e1.Stats = append(e1.Stats, 4)
	e1.Jersey[0] = 17

	// print
	fmt.Println(e1) // {1 ohtani [1 2 3 4] [17]}
	fmt.Println(e2) // {3 freeman [3 4 5] [5]}
	fmt.Println(e3) // {3 freeman [1 2 3] [5]} - no 4
}

File: memory/copy/struct.rs
// rustc struct.rs 
// .\struct.exe

#[derive(Clone, Debug)]
struct Employee {
    id: i32,
    name: String, // immutable utf-8 encoded
    stats: Vec<i32>,
    jersey: [i32; 1],
}

// Function to perform a deep copy of the Employee struct
fn deep_copy(e: &Employee) -> Employee {
    Employee {
        id: e.id,
        name: e.name.clone(), // string as immutable, shallow copy is fine
        stats: e.stats.clone(), // deep copy of the stats vector
        jersey: e.jersey, // array is copied by value
    }
}

fn main() {
    // Initialize struct
    let mut e1 = Employee {
        id: 3,
        name: "freeman".to_string(), // just a pointer to string
        stats: vec![3, 4, 5], // deep copied with clone
        jersey: [5],
    };

    // deep copy
    let e2 = deep_copy(&e1);

    // shallow copy - is deep for strings and vectors
    let e3 = e1.clone();

    // Update
    e1.id = 1;
    e1.name = "ohtani".to_string();
    e1.stats[0] = 1;
    e1.stats[1] = 2;
    e1.stats[2] = 3;
    e1.stats.push(4);
    e1.jersey[0] = 17;

    // Print
    println!("{:?}", e1); // Employee { id: 1, name: ohtani, stats: [1, 2, 3, 4], jersey: [17] }
    println!("{:?}", e2); // Employee { id: 3, name: freeman, stats: [3, 4, 5], jersey: [5] }
    println!("{:?}", e3); // Employee { id: 3, name: freeman, stats: [3, 4, 5], jersey: [5] }
}